### DISCUSSION TITLE
[Q] Streaming XSLX from S3 back to S3 as a CSV

### DISCUSSION THREAD

**Discussion by colinroemer** [exceljs/exceljs#2425] [Q&A]
I am tasked with streaming an XSLX file from S3 bucket A, to S3 bucket B but as a CSV.  My initial assumption was to create a readable stream from bucket A passing that data through the workbookReader below.  as the parsed data is emitted as 'row', i would add that row data the the workbookWrite worksheet and commit.  the workbookWriter is provided a stream.passthrough that also calls s3.upload({stream, bucket, key}).  I know this is an odd request and maybe I am going about it the wrong way but any guidance would be greatly appreciated.

`
      const workbookReader = new ExcelJS.stream.xlsx.WorkbookReader(incomingStream, options);
      const workbookWriter = new Excel.Workbook();
      workbookWriter.csv.write(destinationStream);
      const writeWorksheet = workbookWriter.addWorksheet('test');

      workbookReader.on('worksheet', worksheet => {
        worksheet.on('row', row => {
          writeWorksheet.addRow(row).commit();
        });
      });

      workbookReader.on('end', () => {

        done();
      });

      workbookReader.read();
`

**Comment by mitul-srashtasoft** [comment]
**This is might be helpful to you.**
    
`var params = {
            Bucket: bucketName,
            Key: "a6840015_68f6_402d_a8ad_9d1076a7d925"
        };
    
        var file = s3obj.getObject(params).createReadStream();
        var buffers = [];
    
        file.on('data', function (data) {
            buffers.push(data);
        });
    
        file.on('end', async function () {
            var buffer = Buffer.concat(buffers);
            const stream: Buffer = await this.createExcelFile( buffer );
            res.setHeader( 'Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' );
            res.setHeader( 'Content-Disposition', `attachment; filename=${'TagReport'}.xlsx` );
            res.setHeader( 'Content-Length', stream.length );
            res.send( stream );
            
        });`

` const workbook = new excel.Workbook()
        workbook.xlsx.write(buffer)
        

        await workbook.xlsx.load(buffer)
        
        return new Promise( ( resolve, reject ): void => {
            workbook.commit().then( () => {
                const stream: any = ( workbook as any ).stream;
                const result: Buffer = stream.read();
                resolve( result );
            } ).catch( ( e ) => {
                reject( e );
            } );
        } );`

**Comment by MichaelWalker-git** [comment]
```
    const Bucket = event.Records[0].s3.bucket.name;
    const Key = decodeURIComponent(event.Records[0].s3.object.key.replace(/\+/g, ' '));
    // Retrieve the object
    const data = await s3.getObject({ Bucket, Key }).createReadStream();

    const workbookReader = new ExcelJS.stream.xlsx.WorkbookReader(data);
    for await (const worksheetReader of workbookReader) {
        for await (const row of worksheetReader) {
            console.log("Raw text:\n" + row.values);
        }
    }

```

I did this a little differently.
